README.md

[![CRAN_Status_Badge](http://www.r-pkg.org/badges/version/predictrace)](https://cran.r-project.org/package=predictrace)
[![AppVeyor Build
Status](https://ci.appveyor.com/api/projects/status/github/jacobkap/predictrace?branch=master&svg=true)](https://ci.appveyor.com/project/jacobkap/predictrace)
[![Coverage
status](https://codecov.io/gh/jacobkap/predictrace/branch/master/graph/badge.svg)](https://app.codecov.io/github/jacobkap/predictrace?branch=master)
[![](http://cranlogs.r-pkg.org/badges/grand-total/predictrace?color=blue)](https://cran.r-project.org/package=predictrace)

## Overview

The goal of `predictrace` is to predict the race of a surname or first
name and the gender of a first name. This package uses U.S. Census data
which says how many people of each race has a certain surname. For first
name data, this package uses data from Tzioumis (2018). From this we can
predict which race is mostly likely to have that surname or first name.
The possible races are American Indian, Asian, Black, Hispanic, White,
or two or more races. For the gender of first names, this package uses
data from the United States Social Security Administration (SSA) that
tells how many people of a given name are female and how many are male
(no other genders are included). I use this to determine the proportion
of each gender a name is, and use the gender with the higher proportion
as the most likely gender for that name. Please note that the Census
data on the race of first names is far smaller than the SSA data on the
gender of first names, so you will match far fewer first names to race
than to gender.

Full citation for the Tzioumis data: Tzioumis, Konstantinos (2018)
Demographic aspects of first names, Scientific Data, 5:180025
\[dx.doi.org/10.1038/sdata.2018.25\].

## Installation

``` r
To install this package, use the code
install.packages("predictrace")


# The development version is available on Github.
# install.packages("devtools")
devtools::install_github("jacobkap/predictrace")
```

## Usage

``` r
library(predictrace)
```

    ## Thank you for using predictrace!

    ## To acknowledge our work, please cite the package:

    ## Kaplan, J (2023). predictrace: Predict the Race and Gender of a Given Name Using Census and Social Security Administration Data. Version 1.7.1. URL: https://github.com/jacobkap/predictrace, https://jacobkap.github.io/predictrace/.

### Race of a surname

To get the race of a surname, the only required parameter in
`predict_race()` is `name` which is the surname you want to find the
race of. Please note that this parameter only accepts surnames,
including both first and last name will result in not finding a match in
the Census data. This function returns the name you input, the named
matched through Census data, and likely race of that name, and the
probability that the name is of each race.

``` r
predict_race("Washington")
```

    ##         name match_name likely_race probability_american_indian
    ## 1 Washington washington       black                      0.0066
    ##   probability_asian probability_black probability_hispanic probability_white
    ## 1            0.0028            0.8865               0.0202            0.0517
    ##   probability_2races
    ## 1             0.0323

This function accepts a single string or a vector of strings.

``` r
predict_race(c("Washington", "Franklin", "Lincoln"))
```

    ##         name match_name likely_race probability_american_indian
    ## 1 Washington washington       black                      0.0066
    ## 2   Franklin   franklin       white                      0.0085
    ## 3    Lincoln    lincoln       white                      0.0373
    ##   probability_asian probability_black probability_hispanic probability_white
    ## 1            0.0028            0.8865               0.0202            0.0517
    ## 2            0.0050            0.3828               0.0222            0.5577
    ## 3            0.0125            0.1421               0.0215            0.7615
    ##   probability_2races
    ## 1             0.0323
    ## 2             0.0238
    ## 3             0.0250

If you only want the most likely race and not the individual
probabilities of each race, set the parameter `probability` to FALSE.

``` r
predict_race("Washington", probability = FALSE)
```

    ##         name match_name likely_race
    ## 1 Washington washington       black

### Race of a first name

To get the race of a first name, you again use `predict_race()` but now
set the parameter `surname` to FALSE. This returns the exact same
results as above, but now only works for first names (you may still get
a match of a name if you don’t set the parameter to FALSE but that is
simply because some first names are also last names. The race results
will likely be incorrect.).

``` r
predict_race("sarah", surname = FALSE)
```

    ##    name match_name likely_race probability_american_indian probability_asian
    ## 1 sarah      sarah       white                      0.0014            0.0208
    ##   probability_black probability_hispanic probability_white probability_2races
    ## 1            0.0175               0.0189            0.9397             0.0017

This function accepts a single string or a vector of strings.

``` r
predict_race(c("sarah", "jaime", "jon", "bao"), surname = FALSE)
```

    ##    name match_name likely_race probability_american_indian probability_asian
    ## 1 sarah      sarah       white                      0.0014            0.0208
    ## 2 jaime      jaime    hispanic                      0.0013            0.0493
    ## 3   jon        jon       white                      0.0014            0.0190
    ## 4   bao        bao       asian                      0.0090            0.9640
    ##   probability_black probability_hispanic probability_white probability_2races
    ## 1            0.0175               0.0189            0.9397             0.0017
    ## 2            0.0045               0.4933            0.4497             0.0019
    ## 3            0.0073               0.0100            0.9616             0.0007
    ## 4            0.0000               0.0000            0.0180             0.0090

If you only want the most likely race and not the individual
probabilities of each race, set the parameter `probability` to FALSE.

``` r
predict_race("sarah", probability = FALSE, surname = FALSE)
```

    ##    name match_name likely_race
    ## 1 sarah      sarah       white

### Gender of a first name

To get the gender of a first name you use the `predict_gender()`
function and the only required parameter in `predict_gender()` is `name`
which is the first name you want to find the gender of. This function
returns the name you input, the named matched through SSA data, and
likely gender of that name, and the probability that the name is female
or male.

``` r
predict_gender("tyrion")
```

    ##     name match_name likely_gender probability_female probability_male
    ## 1 tyrion     tyrion          male          0.0115894        0.9884106

``` r
predict_gender(c("tyrion", "jaime", "jon", "sansa", "arya", "cersei"))
```

    ##     name match_name likely_gender probability_female probability_male
    ## 1 tyrion     tyrion          male        0.011589404        0.9884106
    ## 2  jaime      jaime          male        0.420397758        0.5796022
    ## 3    jon        jon          male        0.007450144        0.9925499
    ## 4  sansa      sansa        female        1.000000000        0.0000000
    ## 5   arya       arya        female        0.892700886        0.1072991
    ## 6 cersei     cersei        female        1.000000000        0.0000000

If you only want the most likely gender and not the individual
probabilities of each gender, set the parameter `probability` to FALSE.

``` r
predict_gender(c("tyrion", "jaime", "jon", "sansa", "arya", "cersei"), probability = FALSE)
```

    ##     name match_name likely_gender
    ## 1 tyrion     tyrion          male
    ## 2  jaime      jaime          male
    ## 3    jon        jon          male
    ## 4  sansa      sansa        female
    ## 5   arya       arya        female
    ## 6 cersei     cersei        female

In cases where there is no match in the data, it will return NA.

``` r
predict_gender("zzztyrion")
```

    ##        name match_name likely_gender probability_female probability_male
    ## 1 zzztyrion  zzztyrion          <NA>                 NA               NA


NEWS.md
# predictrace 2.0.1

* Adds startup message.

# predictrace 2.0.0

* Adds parameter in `predict_race()` where you can enter a first name and get the 
  likely race of that first name.
* Adds function `predict_gender()` that uses United States Social Security 
  Administration (SSA) data from 1880-2019 to predict the gender of a first
  name.

# predictrace 1.2.1

* Combines 2010 Census data and 2000 Census data together. 

# predictrace 1.1.1

* Fix bug where Hispanic names are classified as White. Closes #1 by @jmrphy. 

# predictrace 1.0.1

* Adds GitHub links on DESCRIPTION.

# predictrace 1.0.0

* Added a `NEWS.md` file to track changes to the package.


tests/testthat.R
library(testthat)
library(predictrace)

test_check("predictrace")


tests/testthat/test-race.R
smith         <- predictrace::predict_race("smith")
smith_lincoln <- predictrace::predict_race(c("smith", "Lincoln"))
ben_smith     <- predictrace::predict_race("Ben smith")
lincoln       <- predictrace::predict_race("Lincoln")
rlincoln      <- predictrace::predict_race("RLincoln")
perez         <- predictrace::predict_race("Perez")
lucky         <- predictrace::predict_race("lucky")


abbey         <- predictrace::predict_race("abbey", surname = FALSE)
anibal_kenny  <- predictrace::predict_race(c("anibal", "KennY"), surname = FALSE)
harry_houdini <- predictrace::predict_race("harry houdini", surname = FALSE)
zzzdoan       <- predictrace::predict_race("ZZZDoan", surname = FALSE)
doan          <- predictrace::predict_race("Doan", surname = FALSE)
cedric        <- predictrace::predict_race("cedric", surname = FALSE)

test_that("predict_race works", {

  expect_named(smith, c("name",
                        "match_name",
                        "likely_race",
                        "probability_american_indian",
                        "probability_asian",
                        "probability_black",
                        "probability_hispanic",
                        "probability_white",
                        "probability_2races"))
  expect_named(predict_race("smith", probability = FALSE), c("name",
                                                     "match_name",
                                                     "likely_race"))

  expect_equal(smith$likely_race, "white")
  expect_equal(perez$likely_race, "hispanic")
  expect_equal(lucky$likely_race, "white")
  expect_equal(predict_race("smith", probability = FALSE)$likely_race, "white")
  expect_equal(smith$probability_american_indian, 0.0087)
  expect_equal(smith$probability_asian, 0.0045)
  expect_equal(smith$probability_black, 0.2267)
  expect_equal(smith$probability_hispanic, 0.0199)
  expect_equal(smith$probability_white, 0.7211)
  expect_equal(smith$probability_2races, 0.0191)


  expect_equal(lucky$probability_american_indian, 0.0061)
  expect_equal(lucky$probability_asian, 0.042)
  expect_equal(lucky$probability_black, 0.4466)
  expect_equal(lucky$probability_hispanic, 0.0231)
  expect_equal(lucky$probability_white, 0.4544)
  expect_equal(lucky$probability_2races, 0.0279)

  expect_equal(smith_lincoln$likely_race, c("white", "white"))
  expect_equal(predict_race(c("smith", "Lincoln"), probability = FALSE)$likely_race,
               c("white", "white"))
  expect_equal(smith_lincoln$probability_american_indian,
               c(0.0087, 0.0373))
  expect_equal(smith_lincoln$probability_asian,
               c(0.0045, 0.0125))
  expect_equal(smith_lincoln$probability_black,
               c(0.2267, 0.1421))
  expect_equal(smith_lincoln$probability_hispanic,
               c(0.0199, 0.0215))
  expect_equal(smith_lincoln$probability_white,
               c(0.7211, 0.7615))
  expect_equal(smith_lincoln$probability_2races,
               c(0.0191, 0.025))

  expect_true(is.na(ben_smith$likely_race))
  expect_true(is.na(predict_race("Ben smith", probability = FALSE)$likely_race))
  expect_true(is.na(ben_smith$probability_american_indian))
  expect_true(is.na(ben_smith$probability_asian))
  expect_true(is.na(ben_smith$probability_black))
  expect_true(is.na(ben_smith$probability_hispanic))
  expect_true(is.na(ben_smith$probability_white))
  expect_true(is.na(ben_smith$probability_2races))

  expect_equal(lincoln$likely_race, "white")
  expect_equal(predict_race("Lincoln", probability = FALSE)$likely_race, "white")
  expect_equal(lincoln$probability_american_indian, 0.0373)
  expect_equal(lincoln$probability_asian, 0.0125)
  expect_equal(lincoln$probability_black, 0.1421)
  expect_equal(lincoln$probability_hispanic, 0.0215)
  expect_equal(lincoln$probability_white, 0.7615)
  expect_equal(lincoln$probability_2races, 0.025)

  expect_true(is.na(rlincoln$likely_race))
  expect_true(is.na(predict_race("RLincoln", probability = FALSE)$likely_race))
  expect_true(is.na(rlincoln$probability_american_indian))
  expect_true(is.na(rlincoln$probability_asian))
  expect_true(is.na(rlincoln$probability_black))
  expect_true(is.na(rlincoln$probability_hispanic))
  expect_true(is.na(rlincoln$probability_white))
  expect_true(is.na(rlincoln$probability_2races))

})


test_that("predict_race for first name works", {


  expect_named(abbey, c("name",
                        "match_name",
                        "likely_race",
                        "probability_american_indian",
                        "probability_asian",
                        "probability_black",
                        "probability_hispanic",
                        "probability_white",
                        "probability_2races"))
  expect_named(predict_race("abbey", probability = FALSE, surname = FALSE), c("name",
                                                             "match_name",
                                                             "likely_race"))
  expect_equal(abbey$likely_race, "white")

  expect_equal(doan$likely_race, "asian")
  expect_equal(predict_race("doan", probability = FALSE)$likely_race, "asian")
  expect_equal(doan$probability_american_indian, 0)
  expect_equal(doan$probability_asian, 0.9667)
  expect_equal(doan$probability_black, 0)
  expect_equal(doan$probability_hispanic, 0)
  expect_equal(doan$probability_white, 0.0333)
  expect_equal(doan$probability_2races, 0)


  expect_equal(abbey$probability_american_indian, 0)
  expect_equal(abbey$probability_asian, 0)
  expect_equal(abbey$probability_black, 0.0351)
  expect_equal(abbey$probability_hispanic, 0)
  expect_equal(abbey$probability_white, 0.9649)
  expect_equal(abbey$probability_2races, 0)

  expect_equal(anibal_kenny$likely_race, c("hispanic", "white"))
  expect_equal(predict_race(c("anibal", "KennY"), probability = FALSE, surname = FALSE)$likely_race,
               c("hispanic", "white"))
  expect_equal(anibal_kenny$probability_american_indian,
               c(0, 0.0029))
  expect_equal(anibal_kenny$probability_asian,
               c(0, .3736))
  expect_equal(anibal_kenny$probability_black,
               c(0, 0.0402))
  expect_equal(anibal_kenny$probability_hispanic,
               c(.8889, 0.0287))
  expect_equal(anibal_kenny$probability_white,
               c(.1111, 0.5488))
  expect_equal(anibal_kenny$probability_2races,
               c(0, 0.0058))

  expect_true(is.na(harry_houdini$likely_race))
  expect_true(is.na(predict_race("harry houdini", probability = FALSE, surname = FALSE)$likely_race))
  expect_true(is.na(harry_houdini$probability_american_indian))
  expect_true(is.na(harry_houdini$probability_asian))
  expect_true(is.na(harry_houdini$probability_black))
  expect_true(is.na(harry_houdini$probability_hispanic))
  expect_true(is.na(harry_houdini$probability_white))
  expect_true(is.na(harry_houdini$probability_2races))

  expect_equal(cedric$likely_race, "black")
  expect_equal(predict_race("cedric", probability = FALSE, surname = FALSE)$likely_race, "black")
  expect_equal(cedric$probability_american_indian, 0)
  expect_equal(cedric$probability_asian, 0.0909)
  expect_equal(cedric$probability_black, 0.6212)
  expect_equal(cedric$probability_hispanic, 0)
  expect_equal(cedric$probability_white, 0.2727)
  expect_equal(cedric$probability_2races, 0.0151)

  expect_true(is.na(zzzdoan$likely_race))
  expect_true(is.na(predict_race("zzzdoan", probability = FALSE, surname = FALSE)$likely_race))
  expect_true(is.na(zzzdoan$probability_american_indian))
  expect_true(is.na(zzzdoan$probability_asian))
  expect_true(is.na(zzzdoan$probability_black))
  expect_true(is.na(zzzdoan$probability_hispanic))
  expect_true(is.na(zzzdoan$probability_white))
  expect_true(is.na(zzzdoan$probability_2races))

})


tests/testthat/test-test-gender.R



abbey        <- predictrace::predict_gender("abbey")
tyrion_jaime <- predictrace::predict_gender(c("tyrion", "JAIME"))
sansa        <- predictrace::predict_gender("sansa")
zzzjonsnow   <- predictrace::predict_gender("zzzjonsnow")
jorah        <- predictrace::predict_gender("jorah")
cedric       <- predictrace::predict_gender("cedric")



test_that("predict_gender for first name works", {

  expect_named(abbey, c("name",
                        "match_name",
                        "likely_gender",
                        "probability_female",
                        "probability_male"))
  expect_named(predict_gender("abbey", probability = FALSE), c("name",
                                                               "match_name",
                                                               "likely_gender"))
  expect_equal(abbey$likely_gender, "female")

  expect_equal(sansa$likely_gender, "female")
  expect_equal(predict_gender("sansa", probability = FALSE)$likely_gender, "female")
  expect_equal(sansa$probability_female, 1)
  expect_equal(sansa$probability_male, 0)


  expect_equal(jorah$probability_female, 0.26708, tolerance = 4)
  expect_equal(jorah$probability_male, 0.73290, tolerance = 4)

  expect_equal(tyrion_jaime$likely_gender, c("male", "male"))
  expect_equal(predict_gender(c("tyrion", "JAIME"), probability = FALSE)$likely_gender,
               c("male", "male"))
  expect_equal(tyrion_jaime$probability_female,
               c(0.011589, 0.420398), tolerance = 4)
  expect_equal(tyrion_jaime$probability_male,
               c(0.988411, 0.579602), tolerance = 4)

  expect_equal(cedric$likely_gender, "male")
  expect_equal(predict_gender("cedric", probability = FALSE)$likely_gender, "male")
  expect_equal(cedric$probability_female, 0.00677, tolerance = 4)
  expect_equal(cedric$probability_male, 0.99323, tolerance = 4)

  expect_true(is.na(zzzjonsnow$likely_gender))
  expect_true(is.na(predict_gender("zzzjonsnow", probability = FALSE)$likely_gender))
  expect_true(is.na(zzzjonsnow$probability_female))
  expect_true(is.na(zzzjonsnow$probability_male))

})


R/data.R
#' Surnames and number of people of each race with that surname.
#'
#' A dataset containing over 167 thousands surnames and the
#' number of people of each race with that surname.
#'
#' @format A data frame with 167,408 rows and 8 variables:
#' \describe{
#'   \item{name}{Surname}
#'   \item{likely_race}{The most likely race based on the probability of each race}
#'   \item{probability_american_indian}{Probability that the surname is American Indian}
#'   \item{probability_asian}{Probability that the surname is Asian}
#'   \item{probability_black}{Probability that the surname is Black}
#'   \item{probability_hispanic}{Probability that the surname is Hispanic}
#'   \item{probability_white}{Probability that the surname is White}
#'   \item{probability_2races}{Probability that the surname is two or more races}
#'   ...
#' }
#' @source \url{https://www.census.gov/topics/population/genealogy/data/2010_surnames.html} \url{https://www.census.gov/topics/population/genealogy/data/2000_surnames.html}
#'
"surnames_race"


#' Surnames and number of people of each race with that first name
#'
#' A dataset containing over 167 thousands surnames and the
#' number of people of each race with that surname. Citation
#' for this data: Tzioumis, Konstantinos (2018) Demographic aspects of first names, Scientific Data, 5:180025 [dx.doi.org/10.1038/sdata.2018.25].
#'
#' @format A data frame with 4,251 rows and 8 variables:
#' \describe{
#'   \item{name}{Surname}
#'   \item{likely_race}{The most likely race based on the probability of each race}
#'   \item{probability_american_indian}{Probability that the surname is American Indian}
#'   \item{probability_asian}{Probability that the surname is Asian}
#'   \item{probability_black}{Probability that the surname is Black}
#'   \item{probability_hispanic}{Probability that the surname is Hispanic}
#'   \item{probability_white}{Probability that the surname is White}
#'   \item{probability_2races}{Probability that the surname is two or more races}
#'   ...
#' }
#' @source \url{https://dataverse.harvard.edu/dataset.xhtml?persistentId=doi:10.7910/DVN/TYJKEZ}
#'
"first_names_race"


#' Surnames and number of people of each race with that first name
#'
#' A dataset containing almost 100,000 first names and the proportion of people
#' with that first name that are female and male.
#'
#' @format A data frame with 99,444 rows and 4 variables:
#' \describe{
#'   \item{name}{The person's first name}
#'   \item{probability_male}{Probability that the first is male}
#'   \item{probability_female}{Probability that the first name is female}
#'   \item{likely_gender}{The most likely gender based on the probability of each gender}
#'   ...
#' }
#' @source \url{https://www.ssa.gov/oact/babynames/limits.html}
#'
"first_names_gender"


R/hello.R
#' Find the gender of a name
#'
#' @param names
#' Name that you want to know the sex of
#' @param probability
#' The probability that the name is male or female
#'
#' @return
#' A dataframe with three columns: The first column has the name as inputted,
#' the second column has the cleaned up name (no spaces or punctuation,
#' first letter is capitalized), the final column tells the sex of the name
#' @export
#'
#' @examples
#' find_sex("Benjamin")
#' find_sex("Benjamin", probability = TRUE)
#'
#' find_sex(c("Benjamin", "Ge orge", "thom-as", "son-ya", "elena", "RUTH",
#'            "Ivory", "Jessie"))
#' find_sex(c("Benjamin", "Ge orge", "thom-as", "son-ya", "elena", "RUTH",
#'            "Ivory", "Jessie"),  probability = TRUE)
#'
#' example <- data.frame(names =
#'         c("Benjamin", "Ge orge", "thom-as", "son-ya", "elena", "RUTH"))
#' find_sex(example$names)
find_sex <- function(names, probability = FALSE) {

  simpleCap <- function(x) {
    s <- strsplit(x, " ")[[1]]
    paste(toupper(substring(s, 1,1)), substring(s, 2),
          sep = "", collapse = " ")
  }

  names2 <- tolower(names)
  names2 <- gsub("[[:punct:]]", "", names2)
  names2 <- gsub(" ", "", names2)

  return_frame <- data.frame(names)
  return_frame$clean_names <- names2
  return_frame$clean_names <- sapply(return_frame$clean_names,
                                     simpleCap)
  return_frame$sex <- NA

  subset_first <- gender_names[gender_names$name == names2[1],][1,]
  if (is.na(subset_first$male_prob)) {
    return_frame$sex[1] <- NA
  } else if (subset_first$male_prob > subset_first$female_prob) {
    return_frame$sex[1] <- "Male"
    if (probability) {
      return_frame$probability[1] <- subset_first$male_prob
    }
  } else if (subset_first$male_prob < subset_first$female_prob) {
    return_frame$sex[1] <- "Female"
    if (probability) {
      return_frame$probability[1] <- subset_first$female_prob
    }
  }

  if (length(names) > 1) {
    for (i in 2:length(names)) {
      subset_first <- gender_names[gender_names$name == names2[i],][1,]
      if (is.na(subset_first$male_prob)) {
        return_frame$sex[i] <- NA
      } else if (subset_first$male_prob > subset_first$female_prob) {
        return_frame$sex[i] <- "Male"
        if (probability) {
          return_frame$probability[i] <- subset_first$male_prob
        }
      } else if (subset_first$male_prob < subset_first$female_prob) {
        return_frame$sex[i] <- "Female"
        if (probability) {
          return_frame$probability[i] <- subset_first$female_prob
        }
      }
    }
  }

  return(return_frame)
}


#' Find the race of a surname
#'
#' @param surnames
#' Surname that you want to know the race of
#'
#' @param probability
#' The probability that the surname is the selected race
#'
#' @return
#' A dataframe with three columns: The first column has the name as inputted,
#' the second column has the cleaned up name (no spaces or punctuation,
#' first letter is capitalized), the final column tells the race of the name
#' @export
#'
#' @examples
#' find_race("franklin")
#' find_race("franklin", probability = TRUE)
#'
#' find_race(c("franklin", "Washington", "Jefferson", "Sotomayor", "Liu"))
#'
#' find_race(c("franklin", "Washington", "Jefferson", "Sotomayor", "Liu"),
#'              probability = TRUE)
#' example <- data.frame(surnames = c("frank'.lin",
#'                                    "washing-ton", "JEFFERSON", "A dams",
#'                                    "Liu", "Sotomayor"))
#' find_race(example$surnames)
find_race <- function(surnames, probability = FALSE, all = FALSE) {

  simpleCap <- function(x) {
    s <- strsplit(x, " ")[[1]]
    paste(toupper(substring(s, 1,1)), substring(s, 2),
          sep = "", collapse = " ")
  }


  surnames2 <- tolower(surnames)
  surnames2 <- gsub("[[:punct:]]", "", surnames2)
  surnames2 <- gsub(" ", "", surnames2)
  return_frame <- data.frame(surnames)
  return_frame$clean_surnames <- surnames2
  return_frame$clean_surnames <- sapply(return_frame$clean_surnames,
                                        simpleCap)
  return_frame$race <- NA


  subset_first <- surnames_race[surnames_race$name == surnames2[1],][1,]
  max_value <- names(sort(apply(subset_first[2:6], 2,
                                function(x) max(x, na.rm = FALSE)),
                          decreasing = TRUE)[1])
  if (is.na(max_value)) {
    return_frame$race[1] <- NA
  } else (return_frame$race[1] <- gsub("(.*)_percent", "\\1", max_value))
  if (probability) {
    return_frame$probability[1] <- as.numeric(
           sort(apply(subset_first[2:6], 2,
                function(x) max(x, na.rm = FALSE)),
                decreasing = TRUE)[1])
  }

  if (length(surnames) > 1) {
    for (i in 2:length(surnames)) {
      subset_first <- surnames_race[surnames_race$name == surnames2[i],][1,]
      max_value <- names(sort(apply(subset_first[2:6], 2,
                                    function(x) max(x, na.rm = FALSE)),
                              decreasing = TRUE)[1])
      if (is.na(max_value)) {
        return_frame$race[i] <- NA
      } else (return_frame$race[i] <- gsub("(.*)_percent", "\\1", max_value))
      if (probability) {
        return_frame$probability[i] <- as.numeric(
          sort(apply(subset_first[2:6], 2,
                     function(x) max(x, na.rm = FALSE)),
               decreasing = TRUE)[1])
      }
    }
  }

  return(return_frame)
}


R/main.R
#' Find the race of a surname or first name
#'
#' The surname data comes from the United States Census. The first name
#' data comes from Tzioumis (2018, <dx.doi.org/10.1038/sdata.2018.25>)
#'
#' @param name
#' String or vector of strings of surname or first name that you want to know the race of.
#' @param probability
#' If TRUE (default) will provide columns for each race with the probability
#' that the surname is of that race. If FALSE, will only return the name,
#' the match-name from the Census data, and the most likely race.
#' @param surname
#' If TRUE (default) will return the race based on the inputted name being a surname.
#' If FALSE, will return the race based on the inputted name being a first name.
#'
#' @return
#' A data.frame with three or nine columns: The first column has the name as inputted,
#' the second column has the cleaned up name (no spaces or punctuation, all
#' lowercase), the third column tells the likely race of the surname or first name
#' (if there are multiple races with the same probability of a match, it will be
#' a string with each race separated by a comma). If the
#' parameter probability is false, these three columns are all that is returned.
#' Otherwise, columns 4-9 tell the specific probability that the surname or first name
#' is each race.
#' @export
#'
#' @examples
#' predict_race("franklin")
#'
#' predict_race(c("franklin", "Washington", "Jefferson", "Sotomayor", "Liu"))
#' predict_race("franklin", probability = FALSE)
#' predict_race("jacob", probability = FALSE, surname = FALSE)
#' predict_race("jacob", probability = TRUE, surname = FALSE)
predict_race <- function(name, probability = TRUE, surname = TRUE) {

  if (!is.character(name)) {
    stop("name must be a character type.")
  }

  if (is.numeric(probability) ||
      length(probability) != 1 ||
      !probability %in% c(TRUE, FALSE)) {
    stop("probability must either be TRUE or FALSE.")
  }

  data <- data.frame(old_name = name,
                     name     = name)

  data$name <- tolower(data$name)
  data$name <- gsub("[[:punct:]]| ", "", data$name)
  if (surname) {
    data <- dplyr::left_join(data, predictrace::surnames_race, by = "name")
  } else {
    data <- dplyr::left_join(data, predictrace::first_names_race, by = "name")
  }


  names(data) <- gsub("^name$", "match_name", names(data))
  names(data) <- gsub("^old_name$", "name", names(data))
  if (probability == FALSE) {
    data <- data[, c("name",
                     "match_name",
                     "likely_race")]
  } else {
    data <- data[, c("name",
                     "match_name",
                     "likely_race",
                     "probability_american_indian",
                     "probability_asian",
                     "probability_black",
                     "probability_hispanic",
                     "probability_white",
                     "probability_2races")]
  }
  data <- data.frame(data, stringsAsFactors = FALSE)
  data$name <- as.character(data$name)
  return(data)
}




#' Find the gender of a first name
#'
#' The surname data comes from the United States Social Security Administration (SSA).
#' This data has the number of people with that name that are identified as female
#' or male so the probability female/male is the proportion of all people with that
#' name that are female/male. SSA data is available annually from 1880-2019, this
#' aggregates all years together.
#'
#' @param name
#' String or vector of strings of  the first name that you want to know the gender of.
#' @param probability
#' If TRUE (default) will provide columns for each race with the probability
#' that the first name is of that gender If FALSE, will only return the name,
#' the match-name from the SSA data, and the most likely gender.
#'
#' @return
#' A data.frame with three or nine columns: The first column has the name as inputted,
#' the second column has the cleaned up name (no spaces or punctuation, all
#' lowercase), the third column tells the likely gender of the first name
#' (if there are multiple genders with the same probability of a match, it will be
#' a string with each race separated by a comma). If the
#' parameter probability is false, these three columns are all that is returned.
#' Otherwise, columns 4-5 tell the specific probability that the surname is female
#' or male.
#' @export
#'
#' @examples
#' predict_gender("tyrion")
#'
#' predict_gender(c("harry", "ron", "hermione", "DEAN", "NEVILLE", "Cho"))
#' predict_gender("franklin", probability = FALSE)
#' predict_gender("jacob", probability = FALSE)
#' predict_gender("jacob", probability = TRUE)
predict_gender <- function(name, probability = TRUE) {

  if (!is.character(name)) {
    stop("name must be a character type.")
  }

  if (is.numeric(probability) ||
      length(probability) != 1 ||
      !probability %in% c(TRUE, FALSE)) {
    stop("probability must either be TRUE or FALSE.")
  }

  data <- data.frame(old_name = name,
                     name     = name)

  data$name <- tolower(data$name)
  data$name <- gsub("[[:punct:]]| ", "", data$name)
  data <- dplyr::left_join(data, predictrace::first_names_gender, by = "name")


  names(data) <- gsub("^name$", "match_name", names(data))
  names(data) <- gsub("^old_name$", "name", names(data))
  if (probability == FALSE) {
    data <- data[, c("name",
                     "match_name",
                     "likely_gender")]
  } else {
    data <- data[, c("name",
                     "match_name",
                     "likely_gender",
                     "probability_female",
                     "probability_male")]
  }
  data      <- data.frame(data, stringsAsFactors = FALSE)
  data$name <- as.character(data$name)
  return(data)
}

.onAttach <- function(libname, pkgname) {
  package_citation <- "Kaplan, J (2023). predictrace: Predict the Race and Gender of a Given Name Using Census and Social Security Administration Data. Version 2.0.1. URL: https://github.com/jacobkap/predictrace, https://jacobkap.github.io/predictrace/."
  packageStartupMessage("Thank you for using predictrace!")
  packageStartupMessage("To acknowledge our work, please cite the package:")
  packageStartupMessage(package_citation)
}

